scipy/special/cephes/stirling.c

```diff
@@ -64,50 +63,73 @@ long stirling2(int n, int k){
      * to bottom and then left to right. See
      * https://github.com/scipy/scipy/pull/18103#discussion_r1130036365
      */
-    for (int i = 0; i < arraySize; i++){
-        curr[i] = 1L;
+    for (uint64_t i = 0; i < arraySize; i++){
+        curr[i] = 1;
     }
-    long tmp;
+    uint64_t tmp;
     if (k <= n - k + 1) {
-        for (int i = 1; i < n - k + 1; i++){
-            for (int j = 1; j < k; j++){
-	      if (curr[j] > LONG_MAX / (j + 1)) {
+        for (uint64_t i = 1; i < n - k + 1; i++){
+            for (uint64_t j = 1; j < k; j++){
+	      if (curr[j] > UINT64_MAX / (j + 1)) {
 		free(curr);
 		sf_error("stirling2", SF_ERROR_NO_RESULT,
 		         "integer overflow has occured");
-		return -1L;
+		return 0;
 	      }
 	      tmp = curr[j] * (j + 1);
-	      if (tmp > LONG_MAX - curr[j - 1]) {
+	      if (tmp > UINT64_MAX - curr[j - 1]) {
 		free(curr);
 		sf_error("stirling2", SF_ERROR_NO_RESULT,
 		         "integer overflow has occured");
-		return -1L;
+		return 0;
 	      }
 	      curr[j] = tmp + curr[j - 1];
             }
         }
     } else {
-        for (int i = 1; i < k; i++){
-            for (int j = 1; j < n - k + 1; j++){
-	      if (curr[j - 1] > LONG_MAX / (i + 1)){
+        for (uint64_t i = 1; i < k; i++){
+            for (uint64_t j = 1; j < n - k + 1; j++){
+	      if (curr[j - 1] > UINT64_MAX / (i + 1)){
 		free(curr);
 		sf_error("stirling2", SF_ERROR_NO_RESULT,
 		         "integer overflow has occured");
-		return -1L;
+		return 0;
 	      }
 	      tmp = curr[j - 1] * (i + 1);
-	      if (tmp > LONG_MAX - curr[j]) {
+	      if (tmp > UINT64_MAX - curr[j]) {
 		free(curr);
 		sf_error("stirling2", SF_ERROR_NO_RESULT, 
 		         "integer overflow has occured");
-		return -1L;
+		return 0;
 	      }
 	      curr[j] = tmp + curr[j];
           }
       }
   }
-    long output = curr[arraySize - 1];
+    uint64_t output = curr[arraySize - 1];
     free(curr);
     return output;
 }
+
+
+double stirling2_approx(long n, long k) {
+    uint64_t result;
+    if (n == 0 && k == 0){
+        return 1.0;
+    }
+    if (k <= 0 || k > n || n <= 0){
```

Comment:
```suggestion
    if (k < 0 || k > n || n < 0){
```
because of the added check above (n==0 && k == 0)